### PR TITLE
Add IoU loss for box regression.

### PR DESCRIPTION
**Pull request by mingxingtan**
Partially based on https://github.com/tensorflow/addons/pull/914
Related issue: https://github.com/google/automl/issues/270

### CODE REVIEW

**Comment by mingxingtan** [comment]
@fsx950223 Could you help review this PR? Thanks!

**Comment by fsx950223** [comment]
Since I'm denied. I push the code to my [branch](https://github.com/fsx950223/automl/tree/iou).

**Comment by mingxingtan** [comment]
Thank you @fsx950223.  After you accept the invitation, I believe you are part of this repo's developer now.  You should be able to manage all issues and PRs. Let me know if you still have issues.

**Comment by fsx950223** [comment]
remote: Permission to google/automl.git denied to fsx950223.
fatal: unable to access 'https://github.com/google/automl.git/': The requested URL returned error: 403

I got above error

**Comment by fsx950223** [comment]
> Thank you @fsx950223. After you accept the invitation, I believe you are part of this repo's developer now. You should be able to manage all issues and PRs. Let me know if you still have issues.

I wonder whether I'm added to the repo.

**Comment by mingxingtan** [comment]
@googlebot I consent

**Comment by mingxingtan** [comment]
Hi @fsx950223, due to internal policy, there are still some restrictions on directly writing to this repo, but you should be able to review/comment/open/close PRs or issues.

File: efficientdet/iou_utils.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+  target_boxes = tf.cast(target_boxes, pred_boxes.dtype)
+
+  # t_ denotes target boxes and p_ denotes predicted boxes: (y, x, y_max, x_max)
+  pred_boxes_list = tf.unstack(pred_boxes, None, axis=-1)
+  target_boxes_list = tf.unstack(target_boxes, None, axis=-1)
+  assert len(pred_boxes_list) == len(target_boxes_list)
+  assert len(pred_boxes_list) % 4 == 0
+
+  iou_loss_list = []
+  for i in range(0, len(pred_boxes_list), 4):
```
Comment (fsx950223) [inline efficientdet/iou_utils.py:171]: Could use matrix to calculate iou?

Comment (fsx950223) [inline efficientdet/iou_utils.py:171]: Reshape to [N,9,4], then use matrix calculate?

Comment (mingxingtan) [reply efficientdet/iou_utils.py:171]: Good idea. I will first check this in and factorize it with matrix ops later (since this is not part of the inference graph, the slowness might be acceptable).

BTW, since I submitted the change with your email/name, it also needs your consent. Could you help reply "@googlebot I consent"? Thanks!

**Comment by fsx950223** [comment]
@googlebot I consent

**Comment by mingxingtan** [comment]
Hi @fsx950223 I want to try some of these iou losses. Since you have gotten some gains before, I wonder do you balance the three losses:  classification loss, box l2 loss, and iou loss?

total_loss = class_loss + box_loss_weight * box_loss  + iou_loss_weight * iou_loss

Should the box_loss_weight be zero if we are using iou_loss?

**Comment by fsx950223** [comment]
Yes. I replace MSE with giou loss before.

**Comment by mingxingtan** [comment]
Gotcha, how do you set iou_loss_weight? Just use 1.0?

**Comment by fsx950223** [comment]
> Gotcha, how do you set iou_loss_weight? Just use 1.0?

I haven't test the iou losses in this repo. I set 1.0 in my project.
